### PR TITLE
Port the 3 ppx's to `ppxlib`

### PR DESCRIPTION
**Pull request by pitag-ha**
Before, the 2 ppx rewriters `tyxml-ppx` and `tyxml-jsx` as well as the internal `.ml`-file generator in /syntax/ were based on `omp`: the former two on the `omp` driver and all of them on `Ast_408`. None of those will exist in `omp 2.0.0` anymore. For more information on that `omp` change, see https://discuss.ocaml.org/t/ocaml-migrate-parsetree-2-0-0/5991 by @jeremiedimino.

This PR ports those 3 ppx's from the direct use of `omp` and tools like `ppx_tools_versioned` to `ppxlib`. It also adds a couple of small additions to the `tyxml-ppx`-tests. With this change, compiler-versions >= 4.04 are supported; before it was >= 4.02.

Closes https://github.com/ocaml-ppx/ppxlib/issues/146.

@NathanReb , would you mind having a look at this?

### CODE REVIEW

File: test/test_ppx.ml
Comment (NathanReb) [reply test/test_ppx.ml:11]: I think this is just testing a feature of the ppx that was previously untested. I remember discussing it with @pitag-ha back then.

Comment (Drup) [reply test/test_ppx.ml:11]: Could you be more specific ? Because that's a missfeature. The ppx use reflection to mirror the compile-time rules to the typing rules. If you rename elements like that, you break that underlying assumption.

If you replace the function, you are supposed to replace them by something with the same type.

Comment (NathanReb) [reply test/test_ppx.ml:11]: Ok so just summarize, you don't have a problem with the test per-se but you'd prefer if it redefined a sensible `p` function?

Comment (NathanReb) [reply test/test_ppx.ml:11]: Or to rephrase you'd prefer if it illustrated a proper use of of the `[%html.<module name> ...]` feature.

I'm not sure I understand what you think is a missfeature here.

Comment (Drup) [reply test/test_ppx.ml:11]: Well, for a start, I'm wondering what it's supposed to even test. The property I described is a requirement for the ppx to function. It's not a feature.

Comment (Drup) [reply test/test_ppx.ml:11]: Ah, if your goal is to test replacing the *module* (importantly, this module is supposed to respect the same signature), then sure, but we already do that.

File: syntax/reflect/reflect.ml
Comment (pitag-ha) [reply syntax/reflect/reflect.ml:483]: In ppxlib, that part of the Location is handled differently (in particular, `Location.input_name` doesn't exist anymore). I haven't tested the error locations manually. I'll ask the rest of the ppxlib team about this.

File: test/test_ppx.ml
Comment (pitag-ha) [reply test/test_ppx.ml:11]: About the objective of the test: yes, it's to test replacing the module. I've tried to make that clear with the commit message "Add tyxml-ppx tests: (...) open modules", but maybe the term "replace" would have been better here than "open". Do you want me to change that?

About respecting the module signature: Are you referring to respecting the signature of `HtmlWrapped`? If so, I think I do that, because `p` is a value in `HtmlWrapped` of the same type as `a`. So my intention was to shadow the definition of `p` with another definition of the same type.

Which are the tests you're referring to that already test replacing the module? Before writing this test, my changes already passed the tests without taking the replacing module feature into account. That's why I added it: I just wanted to make sure I wouldn't break that feature when integrating it with ppxlib. I added the test to the PR, because I thought  it might also be helpful for others. But if you don't agree, I can get rid of it.

**Comment by pitag-ha** [comment]
What are you referring to concretely when asking if ppxlib is already thinned down? Do you mean if the base dependency is already dropped? If so: yes, in master that's already done.

I don't know how much work it would be to make ppxlib compatible with 4.03, but I imagine it would be quite a bit of work. I'll ask the rest of the team. What's the reason why you'd like to keep that specific version? 

Your overall feedback "the code looks okay-ish" has given me the impression that you're not too happy with the overall OCaml code quality. If you point me to concrete parts that you find worth improving, I'm happy to improve them and happy to learn.

File: test/test_ppx.ml
Comment (Drup) [reply test/test_ppx.ml:11]: In `test_ppx` we use both `Html` and `HtmlWrapped`, so the fact that we can substitute the module is already tested, but it might only test in one direction. I'm curious how your changes made the test pass while breaking that feature. 

`p` and `a` do not have the same types, so that substitution isn't valid. For something simple that covers the same need, I would prefer `let a ~a (elt : Html_types.a_content_fun list) = Html.a ~a (txt "a link : " :: elt)`, which is always correct, and of the same type.

**Comment by Drup** [comment]
> What are you referring to concretely when asking if ppxlib is already thinned down? Do you mean if the base dependency is already dropped? If so: yes, in master that's already done.

That is indeed what I meant.

> I don't know how much work it would be to make ppxlib compatible with 4.03, but I imagine it would be quite a bit of work. I'll ask the rest of the team. What's the reason why you'd like to keep that specific version?

I prefer to keep compatibility when I can, it's not a very strong requirement for such old versions.

> Your overall feedback "the code looks okay-ish" has given me the impression that you're not too happy with the overall OCaml code quality. If you point me to concrete parts that you find worth improving, I'm happy to improve them and happy to learn.

The code isn't bad at all, just not necessarily written according to my style, and since I'm the most likely to maintain it, I prefer if it fits the way I like. I'll simply rewrite bits to my preferences. I'll take a more thorough look once we merge this when the new ppxlib get released.

File: test/test_ppx.ml
Comment (pitag-ha) [reply test/test_ppx.ml:11]: > I'm curious how your changes made the test pass while breaking that feature.

I first ported the rest ignoring that feature by always passing `None` to `get_modname` which (when implementing the feature) should only be done if there's no module in the extension point name. I was expecting the tests to fail, but they passed.



> For something simple that covers the same need, I would prefer `let a ~a (elt : Html_types.a_content_fun list) = Html.a ~a (txt "a link : " :: elt)`

That gives me the type error "Error: This expression has type string but an expression was expected of type
         string wrap". Do you know how to solve it?

File: syntax/reflect/reflect.ml
Comment (pitag-ha) [reply syntax/reflect/reflect.ml:483]: Thanks for pointing me to that part of the code! I think that that particular line is not necessary, but there was another problem I hadn't noticed. I've fixed that now. 

To test that everything is correct now, Ive provoked an exception by introducing an error into one of the `.mli`-files the `.ml`-files get generated from with this PPX. The location gets reported correctly now (including the `.mli`-filename). The locations inside the generated file will get picked up by the compiler when parsing it. Does that seem convincing?

File: ppx/dune
Comment (pitag-ha) [reply ppx/dune:14]: Ok, I've done that now. I've done the same in syntax/dune, syntax/reflect/dune and jsx/dune, since I have the impression that you prefer that style (and in those files, before `Migrate_parsetree` was opened). Is that correct?

**Comment by avsm** [comment]
@Drup would you have time to look at this soon to get it merged and released? I'm seeing a number of problems in opam-repository with universe splits from tyxml and the new ppxlib, and this PR would resolve them.

**Comment by Drup** [comment]
Arg, thanks for reminding me. I have a deadline early next week, I'll work on this after that.

File: syntax/reflect/reflect.ml
Comment (Drup) [reply syntax/reflect/reflect.ml:483]: Yes, that should be pretty good. In any case I realized it's on the reflect, which is not user-facing.

File: test/test_ppx.ml
Comment (Drup) [reply test/test_ppx.ml:11]: I pushed a fix for the test so that it behaves more to my linking, while still testing what you suggested.

File: jsx/tyxml_jsx.ml
Comment (Drup) [reply jsx/tyxml_jsx.ml]: In the end, I moved everything to a nice object-based visitor, it seems to work as expected.

**Review by Drup — APPROVED** [review]
Ok, I have read everything, and made various changes to my linking, which I have pushed here. I do like the fact that it's finally using proper objects to implement open recursion instead of the previous clumsy encoding.

If one of the ppxlib people is willing, I wouldn't mind a "counter-review" to check that what I did is indeed correct wrt ppxlib. Otherwise, I'll merge everything.

Thanks @pitag-ha for doing the first version and @aantron and @avsm for pinging me.

**Comment by Drup** [comment]
I almost forgot: is it now possible to have expect tests with the ppx ? I tried previously (there are various stuff commented in the test directory), but it was much more difficult than it should.

**Comment by Drup** [comment]
(I'm merging now, as this will allow to upgrade my switches to recent ppxlib&co and move forward, but I still welcome answers/reviews! )

**Comment by pitag-ha** [comment]
> is it now possible to have expect tests with the ppx ?

What are you referring to? Do you want to test `tyxml-ppx` with expect tests? Is something in `ppxlib` making that difficult?

**Comment by Drup** [comment]
> What are you referring to? Do you want to test `tyxml-ppx` with expect tests? Is something in `ppxlib` making that difficult?

Yes, I would like to, to test failures and locations in particular. It used to be quite tricky to do with dune. There is an attempt in the test directory that I commented out because it was too unstable. Is it easier now ?

**Comment by pitag-ha** [comment]
Could you use dune cram tests for that? In ppxlib we use cram tests to test failures and locations (see e.g. [here](https://github.com/ocaml-ppx/ppxlib/blob/master/test/driver/standalone_run_as_ppx/run.t#L11-L17)). Or do you need to do something more specific?

**Pull request merged**
